crates/circuit/src/dag_circuit.rs

```diff
@@ -0,0 +1,6141 @@
+            qubits.to_tuple()?.into_any()
+        } else {
+            PyTuple::new_bound(py, [qubits]).into_any()
+        };
+        let key = PyTuple::new_bound(py, &[qubits.unbind(), params_tuple.into_any().unbind()]);
+        calibrations.set_item(key, schedule)?;
+        Ok(())
+    }
+
+    /// Return True if the dag has a calibration defined for the node operation. In this
+    /// case, the operation does not need to be translated to the device basis.
+    fn has_calibration_for(&self, py: Python, node: PyRef<DAGOpNode>) -> PyResult<bool> {
+        if !self
+            .calibrations
+            .contains_key(node.instruction.operation.name())
+        {
+            return Ok(false);
+        }
+        let mut params = Vec::new();
+        for p in &node.instruction.params {
+            if let Param::ParameterExpression(exp) = p {
+                let exp = exp.bind(py);
+                if !exp.getattr(intern!(py, "parameters"))?.is_truthy()? {
+                    let as_py_float = exp.call_method0(intern!(py, "__float__"))?;
+                    params.push(as_py_float.unbind());
+                    continue;
+                }
+            }
+            params.push(p.to_object(py));
+        }
+        let qubits: Vec<BitType> = self
+            .qubits
+            .map_bits(node.instruction.qubits.bind(py).iter())?
+            .map(|bit| bit.0)
+            .collect();
+        let qubits = PyTuple::new_bound(py, qubits);
+        let params = PyTuple::new_bound(py, params);
+        self.calibrations[node.instruction.operation.name()]
+            .bind(py)
+            .contains((qubits, params).to_object(py))
+    }
+
+    /// Remove all operation nodes with the given name.
+    fn remove_all_ops_named(&mut self, opname: &str) {
+        let mut to_remove = Vec::new();
+        for (id, weight) in self.dag.node_references() {
+            if let NodeType::Operation(packed) = &weight {
+                if opname == packed.op.name() {
+                    to_remove.push(id);
+                }
+            }
+        }
+        for node in to_remove {
+            self.remove_op_node(node);
+        }
+    }
+
+    /// Add individual qubit wires.
+    fn add_qubits(&mut self, py: Python, qubits: Vec<Bound<PyAny>>) -> PyResult<()> {
+        for bit in qubits.iter() {
+            if !bit.is_instance(imports::QUBIT.get_bound(py))? {
+                return Err(DAGCircuitError::new_err("not a Qubit instance."));
+            }
+
+            if self.qubits.find(bit).is_some() {
+                return Err(DAGCircuitError::new_err(format!(
+                    "duplicate qubits {}",
+                    bit
+                )));
+            }
+        }
+
+        for bit in qubits.iter() {
+            self.add_qubit_unchecked(py, bit)?;
+        }
+        Ok(())
+    }
+
+    /// Add individual qubit wires.
+    fn add_clbits(&mut self, py: Python, clbits: &Bound<PySequence>) -> PyResult<()> {
```

Comment:
Should this also be `Vec<Bound<PyAny>>` like it is with `add_qubits`? Or should `add_qubits` be `PySequence` like this one? Is there a reason why we take different things here.

Reply:
To my knowledge, these shouldn't be different. I'd check the Git blame to see which was changed most recently. I'd originally done these at the same time, so they probably _used_ to be the same, and either I or Matt updated one of them

Reply:
Also worth noting, the `Vec` is probably nicer for Rust callers, so if we can, that's probably the better choice

Reply:
I changed `add_qubits` in:  882e6e603d8a6a6973e31234afe38e86f20bd67f prior to the big squash. TBH, I'm not sure what my exact intent was there it was basically a month ago.